JongHo/app.js

```diff
@@ -42,6 +41,119 @@ app.post("/userCreate", async (request, response) => {
   response.status(201).json({ message: "userCreated" });
 });
 
+//create a post
+app.post("/postCreate", async (request, response) => {
+  const { title, content, user_id, image_url } = request.body;
+  await myDataSource.query(
+    `INSERT INTO posts (title,content,user_id,image_url) VALUES (?,?,?,?);`,
+    [title, content, user_id, image_url]
+  );
+  response.status(201).json({ message: "postCreated" });
+});
+
+//inquire posts
+app.get("/posts", async (request, response) => {
+  await myDataSource.query(
+    `SELECT 
+        users.id as userId,
+        users.profile_image as userProfileImage,
+        posts.id as postingId, 
+        posts.image_url as postingImageUrl,
+        posts.content as postingContent 
+    FROM users 
+    INNER JOIN posts 
+    ON posts.user_id=users.id;`,
+    (err, results) => {
+      if (err) {
+        console.error(err);
+      } else {
+        response.status(200).json({ data: results });
+      }
+    }
+  );
+});
+
+//inquire specific user's posts
+app.get("/postUser/:userId", async (request, response) => {
+  const userId = request.params.userId;
+  await myDataSource.query(
+    `SELECT
+        u.id AS userId,
+        u.profile_image AS userProfileImage,
+        JSON_ARRAYAGG(
+            JSON_OBJECT("postingId", p.id, "postingImageUrl", p.image_url, "postingContent", p.content)
+                    ) AS postings
+    FROM users u
+    INNER JOIN posts p
+    ON u.id = p.user_id
+    WHERE u.id = ${userId}
+    GROUP BY u.id;
+    `,
+    (err, results) => {
+      if (err) {
+        console.error(err);
+      } else {
+        response.status(200).json({ data: results });
+      }
+    }
+  );
+});
+
+//modify a post
+app.put("/postModify/:postingId", async (request, response) => {
+  const postingId = request.params.postingId;
+  const { content } = request.body;
+  await myDataSource.query(
+    `UPDATE 
+        posts 
+    SET content=? 
+    WHERE posts.id=?;`,
+    [content, postingId]
+  );
+  await myDataSource.query(
+    `SELECT 
+        posts.user_id as userId,
+        users.name as userName,
+        posts.id as postingId,
+        posts.title as postingTitle,
+        posts.content as postingContent
+    from posts 
+    inner join users 
+    on posts.user_id=users.id 
+    where posts.id=${postingId}`,
```

Comment:
여기는 다른 코드와는 다르게 ?로 처리하지 않고 곧바로 string을 주입하셨는데 특별한 이유가 있을까요?

추가로 ?를 사용하는 방식과 위와 같은 방식이 어떠한 차이가 있는지 서칭 한 번 해보시면 좋을 것 같습니다.

Reply:
```
`//inquire posts
app.get("/posts", async (request, response) => {
  await myDataSource.query(
    `SELECT 
        users.id as userId,
        users.profile_image as userProfileImage,
        posts.id as postingId, 
        posts.image_url as postingImageUrl,
        posts.content as postingContent 
    FROM users 
    INNER JOIN posts 
    ON posts.user_id=users.id;`,
    (err, results) => {
      if (err) {
        console.error(err);
      } else {
        response.status(200).json({ data: results });
      }
    }
  );
});`
```

위와 같은 방식으로 **query의 결과값에 대하여 err나 results를 출력하려고** 처음에는 생각했습니다.
그래서 
`SELECT ~~~,[argument,....],(err, results)=>{} `
방식으로 코드를 작성한 후에 코드를 실행하였더니 

```
`return await usedQueryRunner.query(query, parameters); // await is needed here because we are using finally
                                         ^

TypeError: usedQueryRunner.query is not a function`
```

오류가 발생하였고 DataSource.d.ts에서 
query 메소드는 다음과 같이 사용되는 것을 확인했습니다.

` query(query: string, parameters?: any[], queryRunner?: QueryRunner): Promise<any>;`

여기서 저는 처음에 queryRunner parameter에 (err, results)=>{} 함수를 넣는다고 막연하게 생각했는데
DataSource.js에 확인해보니

```
`async query(query, parameters, queryRunner) {
        if (InstanceChecker_1.InstanceChecker.isMongoEntityManager(this.manager))
            throw new error_2.TypeORMError(`Queries aren't supported by MongoDB.`);
        if (queryRunner && queryRunner.isReleased)
            throw new error_1.QueryRunnerProviderAlreadyReleasedError();
        const usedQueryRunner = queryRunner || this.createQueryRunner();
        try {
            return await usedQueryRunner.query(query, parameters); // await is needed here because we are using finally
        }
        finally {
            if (!queryRunner)
                await usedQueryRunner.release();
        }
    }`

```
(err, results)=>{} 가 'queryRunner'가 아니라 'parameters'로 동작하는 것으로 추정하게 되었습니다.

더 이상 찾아보는게 쉽지 않아서 여기서 궁금한게
위의 내용 처럼 (err, results)=>{}을 parameters로서 사용하게 되면 ,
sql query를 진행할 때 
parameters에 들어온 argument를 불러와 사용하는 것뿐만 아니라
어떻게 query의 결괏값을 
argument에 있는 arrow function에 넣어서 
결과를 출력할 수 있는지 여쭤보고 싶습니다.


추가적으로 ?를 사용할 때랑 string을 사용할 때랑 차이는 찾아보겠습니다.